pkg/apis/workloads/v1alpha1/helpers.go

```diff
@@ -41,6 +45,105 @@ func (c *Console) TTLDuration() time.Duration {
 	return time.Duration(*c.Spec.TTLSecondsAfterFinished) * time.Second
 }
 
+// GetDefaultCommandWithArgs returns a concatenated list of command and
+// arguments, if defined on the template
+func (ct *ConsoleTemplate) GetDefaultCommandWithArgs() ([]string, error) {
+	containers := ct.Spec.Template.Spec.Containers
+	if len(containers) == 0 {
+		return []string{}, errors.New("template has no containers defined")
+	}
+
+	return append(containers[0].Command, containers[0].Args...), nil
+}
+
+// GetAuthorisationRuleForCommand returns an authorisation rule that matches
+// the command that a console is being started with, or an error if one does
+// not exist.
+//
+// It does this by iterating through the console template's authorisation rules
+// list until it finds a match, and then falls back to the default
+// authorisation rule if one is defined.
+//
+// The `matchCommandElements` field, within an AuthorisationRule, is an array
+// of matchers, of which there are 3 supported types:
+//
+//   1. `*`  - a wildcard that matches the presence of an element.
+//   2. `**` - a wildcard that matches any number (including 0) of
+//             elements. This can only be used at the end of the array.
+//   3. Any other string of characters, this is used to perform an exact
+//      string match against the current element.
+//
+// The elements of the command array are evaluated in order; any failure to
+// match will result in falling back to the next rule.
+//
+// Examples:
+//
+// | Matcher               | Command                | Matches? |
+// | --------------------- | ---------------------- | -------- |
+// | ["bash"]              | ["bash"]               | Yes      |
+// | ["ls", "*"]           | ["ls", "file"]         | Yes      |
+// | ["ls", "*"]           | ["ls"]                 | No       |
+// | ["echo", "**"]        | ["echo"]               | Yes      |
+// | ["echo", "**"]        | ["echo", "hello"]      | Yes      |
+// | ["echo", "**"]        | ["echo", "hi", "bye" ] | Yes      |
+// | ["echo", "**", "bye"] | ["echo", "hi", "bye" ] | Error    |
+//
+func (ct *ConsoleTemplate) GetAuthorisationRuleForCommand(command []string) (ConsoleAuthorisationRule, error) {
+	for _, rule := range ct.Spec.AuthorisationRules {
+	elements:
+		for i, element := range rule.MatchCommandElements {
+			switch element {
+			case "*":
```

Comment:
Was the first part of this comment supposed to be dropped on the `"**"` or `""` error cases?

Yeah I think you're right about the validating webhook. I considered that but wanted to avoid the additional boilerplate. It does make sense to have though, and will allow us to do other thing such as validate that if you have authorisation rules defined then you must also set a default rule (if you agree that that's sensible).

Reply:
> I think it would also be good to flip this so we continue on successful matches of an element, and the default be a break

So I did have it written this way originally, but it had an extra variable and looked a bit uglier and harder to read. I can see if it's possible to flip back though and keep it clean.

> That way we at worst disallow consoles that are valid in case there is a mistake in the code]

I don't think this is strictly true though? Even if we defaulted to breaking out, we could still have a bad element matcher which erroneously matched when it shouldn't, therefore allowing a console command that we should not?

I think that the test suite, and coverage stats (which is currently at 100%), is the only way to gain confidence that this is doing the right thing.